### PR TITLE
Add serverless, remove no longer working Trac, BuildbotTryNotifier.

### LINKED ISSUE #19: Convert to Py3 and deploy to Azure Functions
Automatically deploying a python HTTP app from GitHub to Vultr/weesp is not easy, so we should move to a SaaS/PaaS.

Look into hosting it using some Docker container / SaaS python service / Azure functions and setup some automatic deployments.

### PR DESCRIPTION
**Pull request by danuker**
Closes #19 

This is a first try at the `serverless` library. I don't know where the output of the logger is, when I run it with `sls offline`.

To be continued in #25.

To run manually:

1. Install serverless NPM dependencies: `npm install`
2. Run the off-cloud version on localhost: `sls offline`

### CODE REVIEW

File: .npmignore [hunk 1]
```diff
@@ -0,0 +1,103 @@
+# Logs
```
Comment (danuker) [inline .npmignore:1]: This file was generated; no need to review.

File: chevah/github_hooks_server/handler.py [hunk 2]
```diff
@@ -8,8 +8,6 @@
 
 
 from chevah.github_hooks_server import log
-from chevah.github_hooks_server.utils.trac import Trac, TracException
```
Comment (danuker) [inline chevah/github_hooks_server/handler.py:11]: I removed all Trac handling.

File: chevah/github_hooks_server/server.py [hunk 3]
```diff
@@ -48,47 +45,44 @@ def __init__(self, message):
...
+def ping(req: func.HttpRequest):
     """
-    Entry point.
+    Simple resource to check that server is up.
     """
-    return ''
+    return func.HttpResponse('pong')
 
 
-@route('/ping',  methods=['GET'])
```
Comment (danuker) [inline chevah/github_hooks_server/server.py:59]: The route definitions are now in `serverless.yml`.

File: chevah/github_hooks_server/server.py [hunk 4]
```diff
@@ -114,32 +108,24 @@ def hook(request, hook_name):
     handle_event(event)
 
 
-@route('/buildmaster',  methods=['POST'])
-def buildmaster(request):
-    """
-    Buildbot integration entry point.
-    """
-    return BuildbotTryNotifier(CONFIGURATION)
```
Comment (danuker) [inline chevah/github_hooks_server/server.py:122]: I removed `BuildbotTryNotifier` as well, since it wouldn't work after the Py3 upgrade, and I believe it is no longer used either.

File: config.toml [hunk 5]
```diff
@@ -3,12 +3,12 @@
...
-        "50.57.128.197/32",
-        "108.171.174.178/32",
-        "50.57.231.61/32",
-        "204.232.175.64/27",
         "192.30.252.0/22",
+        "185.199.108.0/22",
+        "140.82.112.0/20",
+        "143.55.64.0/20",
+        "2a0a:a440::/29",
+        "2606:50c0::/32"
```
Comment (danuker) [inline config.toml:11]: New GitHub URLs were in https://api.github.com/meta. But it is not used right now. Should I remove it completely?

File: serverless.yml [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+# This file is the main config file for your service.
+# It's very minimal at this point and uses default values.
+# You can always add more config options for more control.
+# We've included some commented out config examples here.
+# Just uncomment any of them to get that config option.
+#
+# For full config options, check the docs:
+#    docs.serverless.com
+#
+# Happy Coding!
```
Comment (danuker) [inline serverless.yml:12]: I kept this file close to the [original template file](https://github.com/serverless/serverless-azure-functions#create-a-new-azure-function-app) generated using `sls create`. Should I clean it up instead?

File: setup.py [hunk 7]
```diff
@@ -41,31 +41,23 @@ def run(self):
...
-        # We keep an older version of python is use.
-        'Twisted==15.5.0.chevah1',
-        'github3-py==1.0.0.gitc82e90e',
-        # We don't depended directly on them but github3.py
-        # dependencies are bad.
-        'urllib3',
-        'chardet',
-        'certifi',
-        'idna',
+        'github3.py',
```
Comment (danuker) [inline setup.py:48]: I did not tesst the actual GitHub integration. It needs the secret, and to be deploymed.

File: serverless.yml [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+  region: West Europe
+  runtime: python3.8 # python3.7 or python3.6 also available
+  # os: linux  # linux is the only operating system available for python
+  # prefix: "sample"  # prefix of generated resource name
+  # subscriptionId: A356AC8C-E310-44F4-BF85-C7F29044AF99
+  # stage: dev
+  # type: premium  # premium azure functions
+
+  environment: # these will be created as application settings
+    VARIABLE_FOO: 'foo'
```
Comment (danuker) [inline serverless.yml:31]: The environment variables might need updating.

**Comment by danuker** [comment]
@adiroiban I doubt this is ready for merge, but would appreciate a first look (I added comments on the diff).

File: serverless.yml (hunk 6 above)
Comment (adiroiban) [reply serverless.yml:12]: I think that you can clean it and add a link to the docs for the format of the file.

File: setup.py (hunk 7 above)
Comment (adiroiban) [reply setup.py:48]: We can test it in a separate PR... it's ok to reduce the scope of this PR and just focus and nodejs tooling and deployment

we can fix the actual GH code in a separate PR.

the tests will need refactoring 

the private file is where you store a github API token... i think, it was a local file to prevent leaking the code via the env or process arguments.

there are some end-to-end tests and they are more suited for manual usage and not automated tests

File: setup.py [hunk 9]
```diff
@@ -41,31 +41,23 @@ def run(self):
...
+        'azure-functions',
+        'azure-functions-worker',
         ],
-    extras_require = {
+    extras_require={
```
Comment (adiroiban) [inline setup.py:52]: don't worry too much about formating... we will move this to `setup.cfg` or `pyproject.toml` 

we might want to experiment with https://python-poetry.org/

File: serverless.yml (hunk 8 above)
Comment (adiroiban) [reply serverless.yml:31]: that is ok

File: chevah/github_hooks_server/server.py (hunk 3 above)
Comment (adiroiban) [reply chevah/github_hooks_server/server.py:59]: that is not nice ... but what can we do :)   this is how servless works

File: .gitignore [hunk 10]
```diff
@@ -13,3 +13,7 @@ virtualenv/
 *.egg-info/
 build/
 twistd.log
+
+# NodeJS
+node_modules/
```
Comment (adiroiban) [inline .gitignore:18]: `node_modules` is some kind of `site-packages` from node.

We will see how we integrate with nodeenv ... we might want to have everything inside a single
`venv` directory  where we store the python venv and where we will also have `venv/node_modules`

File: chevah/github_hooks_server/server.py (hunk 4 above)
Comment (adiroiban) [reply chevah/github_hooks_server/server.py:122]: yes... buildmaster is no longer used.

File: setup.py [hunk 11]
```diff
@@ -41,31 +41,23 @@ def run(self):
...
         'dev': [
             'mock',
             'nose',
             'pyflakes',
             'pep8',
```
Comment (adiroiban) [inline setup.py:57]: here we might want to add `nodeenv` and have a script that will call nodenv with a specific node version

File: setup.py [hunk 12]
```diff
@@ -41,31 +41,23 @@ def run(self):
...
             ],
     },
-    test_suite = 'chevah.github_hooks_server.tests',
+    test_suite='chevah.github_hooks_server.tests',
```
Comment (adiroiban) [inline setup.py:60]: this was experiment to check testing features in `setup.py`

You can remove it and look at running the tests with `py.test` framework

**Review by adiroiban — CHANGES_REQUESTED** [review]
Thanks for the update.

It's a good start...but we need a few build updated and documentation updates.

I think that is best to use node via `nodeenv`

I think that we still need to learn about nodejs dev tools

```
virtualenv venv
. venv/bin/activate
nodeenv venv/node
pip install  .

# would be nice to have just npm install -g
# or nodeenv -r node-requirements.txt venv/node/
npm install -g serverless 
serverless plugin install -n serverless-azure-functions
```

For the future, we also need to make sure that all the npm packages that we use are  "cached" at https://github.com/orgs/chevah/packages

Now that I am looking at serverless ... I am not sure it's the best tool.

Maybe is best to just go with the standard azure tools :)
but we are trying :)

File: setup.py [hunk 13]
```diff
@@ -41,31 +41,23 @@ def run(self):
...
-        'Twisted==15.5.0.chevah1',
-        'github3-py==1.0.0.gitc82e90e',
-        # We don't depended directly on them but github3.py
-        # dependencies are bad.
-        'urllib3',
-        'chardet',
-        'certifi',
-        'idna',
+        'github3.py',
+        'azure-functions',
```
Comment (adiroiban) [inline setup.py:49]: it's best to `pin` the exact version of these packages

File: README.rst [hunk 14]
```diff
@@ -14,3 +14,22 @@ It is not / and should not be published on PyPi.
...
+
+
+serverless
+==========
+
+To run offline for testing purposes:
+
+```sh
+npm install  # Required after changing `serverless.yml`.
+sls offline
```
Comment (adiroiban) [inline README.rst:26]: I think that is best to use `npm install -g` from a nodeenv ...and use `serverless` offline. to reduce the confusion with the `sls` command

and I think that the readme should be updated to instruct that `pip install .` should be called before `servereless offline` to get the packages  and prepare the deployment environment

File: package.json [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+  "name": "github-hooks-server",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
+  "license": "ISC",
+  "devDependencies": {
```
Comment (adiroiban) [inline package.json:11]: why is this dev?
this is not needed for the deployment?

File: setup.py (hunk 9 above)
Comment (danuker) [reply setup.py:52]: I added a tentative `pyproject.toml`. You can use it with `poetry shell`, then run what you need.

File: config.toml (hunk 5 above)
Comment (adiroiban) [reply config.toml:11]: We  can remove... btw. we can also remove toml dependency.

This was an experiment to try toml ... but I think that for technical usage yaml is more popular and for simple things .INI is better 

so, for this project I think that is best to just use .INI

File: local.settings.json [hunk 16]
```diff
@@ -0,0 +1 @@
+{"IsEncrypted":false,"Values":{"AzureWebJobsStorage":"UseDevelopmentStorage=true","FUNCTIONS_WORKER_RUNTIME":"python"}}
```
Comment (adiroiban) [inline local.settings.json:1]: is this file needed?

File: setup.py (hunk 9 above)
Comment (adiroiban) [reply setup.py:52]: have you pushed it ? I don't see any pyproject.toml  ... and I think that once you use pyproject.toml you can remove the setup.py  ... or have just `setup()` inside setup.py to have `pip install =e .`

Now that I am looking at poetry I see that it is sponsored by vercel and that vercel has python serverless functions

https://vercel.com/docs/concepts/functions/supported-languages#supported-languages

so maybe we can deploy this to vercel instead of azure :p

File: .gitignore (hunk 10 above)
Comment (danuker) [reply .gitignore:18]: I tried to have everything installed in a virtualenv with `-g`, but it looks like Serverless plugins [don't work from a global install](https://forum.serverless.com/t/plugins-dont-work-when-installed-globally/5874). So we will have the `node_modules` folder if we use it.

File: setup.py (hunk 9 above)
Comment (danuker) [reply setup.py:52]: No, I commented before pushing; sorry. I was experimenting with the venv also; I have pushed now.

I think I would also prefer not to use this library with tons of insecure dependencies. When pushing I got:
```
remote: GitHub found 5 vulnerabilities on chevah/github-hooks-server's default branch (2 critical, 1 high, 2 moderate). To find out more, visit:
remote:      https://github.com/chevah/github-hooks-server/security/dependabot
remote: 
```

But the pros of serverless is we can easily change to [cheaper providers](https://www.serverless.com/framework/docs/providers). Looking at the [Azure intro](https://www.serverless.com/framework/docs/providers/azure/guide/intro/) from there makes me think we don't need the `serverless-azure-functions` plugin.

I like whatever is developer-friendly and not too costly to adapt (to avoid vendor lock-in).

File: setup.py (hunk 11 above)
Comment (danuker) [reply setup.py:57]: > and have a script that will call nodenv with a specific node version

I added `nodeenv` here, and added the "Development" section in the readme, but do you also want a script creating the environment?

File: .gitignore (hunk 10 above)
Comment (adiroiban) [reply .gitignore:21]: ok. no problem. we can live with node_modules ignored in the root... this is life :) 

but when possible we should prefer a nodeenv and -g

File: setup.py (hunk 11 above)
Comment (adiroiban) [reply setup.py:57]: no. I think that you create the env only once. It's ok to have it only in the readme .

File: setup.py (hunk 9 above)
Comment (adiroiban) [reply setup.py:52]: we tried this "technology" 

for huge projects it might make sense to make it easy to switch between providers ... but for us this is not a big deal.

I went with Azure Functions as it had support for python serverless.

I now see that vercel also has support for Python

but it looks like the future is nodejs for serverless ...so in the end we might just embrace JS for all web related things

File: local.settings.json (hunk 16 above)
Comment (danuker) [reply local.settings.json]: Looks like not; I removed it.

File: package.json (hunk 15 above)
Comment (danuker) [reply package.json]: In the repo it is now under `dependencies`, but I noticed something else editing it back to `devDependencies`, but I have to find out what exactly.

File: README.rst (hunk 14 above)
Comment (danuker) [reply README.rst]: I replaced `sls` with `serverless` everywhere. I used `poetry install` instead of `pip install .`, but `pip` should work also.

File: setup.py (hunk 11 above)
Comment (danuker) [reply setup.py:57]: I will look at `nodeenv` options, to see if I can move everything to one `node_modules` folder.

Comment (danuker) [reply setup.py:57]: I could get `serverless-azure-functions` to install globally in the venv, but it won't run from there.

```
Error ---------------------------------------------------
 
  Error: spawn /home/dan/Documents/Work/ProAtria/github-hooks-server/node_modules/.bin/func ENOENT
      at Process.ChildProcess._handle.onexit (node:internal/child_process:282:19)
      at onErrorNT (node:internal/child_process:475:16)
      at processTicksAndRejections (node:internal/process/task_queues:83:21)
 
     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.
```

**Comment by danuker** [comment]
I added `poetry` support and cleaned up the docs some more.

Should I remove `setup.py` now that there is `pyproject.toml`?

I have not added the PublishCommand; there is [no command execution support for poetry](https://github.com/python-poetry/poetry/issues/265). Should I add it to Makefile?

needs-review

**Comment by adiroiban** [comment]
For publish command is ok.


```
[[tool.poetry.source]]
name = "chevah"
default = true
url = "http://deag.chevah.com:10042"

--

poetry publish -r chevah
```


but now I found out  that while we have removed Trac...we still need those user/pass to publish pypi packages.

We should look at a replacement.

I think that this is ok for a first merge.

------------

we can then set a hook only for this repo and run some tests.

Thanks!
